sycl/tools/sycl-xocc/bin/sycl-xocc

```diff
@@ -13,56 +13,94 @@
-# $4 == our .xpirbc input file or the name of our linker output file
-# $5 == temporary file directory to place intermediate .xo files or to retrieve
-#       intermediate .xo files for linking when -l speciifed
+# $2 == path to the clang driver that's executing the shell script (e.g. 
+#        your compiled Clang's build/bin)
+# $3 == The name of the source file without file extensions or temporary 
+#       manglings, mostly used to generate some intermediate file names
+# $4 == The initial input file name, currently a .o file that's actually a .bc 
+#       file in disguise       
+# $5 == The systems temporary directory where we'll deposit some intermediate 
+#       files
+# $6 == The name we wish to give our final output .xcl binary, important that it
+#       is consistent with what the driver expects it to be. 
 
 # TODO: Pass an argument through the driver to set options like XILINX_PLATFORM, 
 # the environment variables are a quick workaround to swap boards. 
-
 usage() { echo xocc_compile: error: $2 >&2; exit $1; }
 
-[ $# -eq 0      ] && usage 1 "no xocc shell or bin"
-[ $# -eq 0      ] && usage 2 "no compilation option specified"
-[ $# -eq 0      ] && usage 3 "no data file"
+if [[ -z "$1" ]]; then
+  usage 1 "no sdx bin directory containing xocc"
+fi
 
-# load in our information file to a shell variable
-DATA=$(<$3)
+if [[ -z "$2" ]]; then
+  usage 2 "no driver path specified"
+fi
 
-if [[ $2 = "-c" ]]; then
-  [ $# -eq 0      ] && usage 4 "no xpirbc input file"
-  [ $# -eq 0      ] && usage 5 "no file directory for output specified"
+if [[ -z "$3" ]]; then
+  usage 3 "no source file name"
+fi
 
-  while IFS= read -r X
-  do
-  if [[ -n "$X" ]]; then
-    $1 --target $XCL_EMULATION_MODE --platform $XILINX_PLATFORM \
-      --save-temps -c -k $X -o $5/$X.xo $4
-  fi
-  done <<< $DATA
-elif [[ $2 = "-l" ]]; then
-  [ $# -eq 0      ] && usage 4 "no output file specified"
-  [ $# -eq 0      ] && usage 5 "no directory containing .xo's to link specified"
+if [[ -z "$4" ]]; then
+  usage 4 "no input file name"
+fi
 
-  LINKER_LIST=""
-  while IFS= read -r X
-  do
-  if [[ -n "$X" ]]; then
-      LINKER_LIST="$5/$X.xo $LINKER_LIST"
-  fi
-  done <<< $DATA
+if [[ -z "$5" ]]; then
+  usage 5 "no temporary directory specified"
+fi
 
-  $1 --target $XCL_EMULATION_MODE --platform $XILINX_PLATFORM -l \
-    --save-temps -o $4 $LINKER_LIST
+if [[ -z "$6" ]]; then
+  usage 6 "no final output file specified"
 fi
+
+# Hardcoded path from the bin directory that xocc is stored in, works for now 
+# but a future TODO might be to allow this to be optionally specified as an 
+# ENV variable 
+XOCC_LIB_SPIR="$1/../lnx64/lib/libspir64-39-hls.bc"
+
+# Perhaps not so important that the linker is the one packaged with the Clang
+# build, but it's important that Opt is so that it's packaged with the correct 
+# LLVM passes. 
+OPT="$2/opt"
+LLVM_LINK="$2/llvm-link"
+XOCC="$1/xocc"
+
+$OPT -O3 -asfix -globaldce -inSPIRation -globaldce -kernelNameGen "$4" -o \
```

Comment:
Should the `-O3` come from a compiler option instead?
And how to provide `-g` or whatever other option?

Reply:
I think this should come in a separate patch, there are a lot of things desirable to be passed through to the script e.g. XILINX_PLATFORM and XCL_EMULATION_MODE would be better off as arguments than environment variables. However, at the bare minimum it requires parsing and then passing through more command line arguments via the XOCC ToolChain as well as some script modifications. It would also be interesting to look at some of the other possible compiler options we could filter down.

Perhaps a little outside of the scope of this patch? And can be opened up as a new issue for a future patch?